aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+
+#include <ATen/Context.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
+#include <ATen/native/DispatchStub.h>
+#include <c10/core/ScalarType.h>
+#include <c10/util/env.h>
+#include <c10/util/irange.h>
+
+#include <functional>
+#include <unordered_set>
+#include <vector>
+#include <ATen/cuda/CUDAContext.h>
+
+namespace sdp {
+
+struct sdp_params {
+  const at::Tensor& query;
+  const at::Tensor& key;
+  const at::Tensor& value;
+  bool has_attn_mask;
+  double dropout;
+  bool need_attn_weights;
+  bool is_causal;
+};
+
+enum class SDPBackend {flash_attention, math, error};
+
+//  Define gate functions that determine if a flash kernel can be ran
+inline bool check_gradients(sdp_params params, bool debug) {
```

Comment:
We already check for this at the SDP top level so this should be safe to remove, I want to be as restrictive as possible initially but this should be removed